### PR TITLE
ISSUE #3248 Support for ALTER TABLE altering columns NOT NULL

### PR DESCRIPTION
**Pull request by zippond**
**Desc.**
This PR implement the "ALTER" part for the column "NOT NULL" constraint.
"SET NOT NULL" will failed if column contains NULL value.
"DROP/SET NOT NULL" can be repeatedly execute against a specific column. 

I have some questions when working on this feature. So this might be a discussion rather than a PR.

**Questions.**
1. In TableCatalogEntry::SetNotNull(), I called DataTable::Scan() with a IsNullFilter() to check if the column contains NULL values. 
Is it the right way/place to do so?
2. As I pushed a IsNullFilter to TableFilterSet, the RowGroup::CheckZonemap() is called during the scan. It seems all the columns should be pushed back to column_ids. Did I missed anything? 
```
	235 bool RowGroup::CheckZonemap(TableFilterSet &filters, const vector<column_t> &column_ids) {
	236     for (auto &entry : filters.filters) {
	237         auto column_index = entry.first;
	238         auto &filter = entry.second;
	239         auto base_column_index = column_ids[column_index];	--> If column_index is the last column, all clumn ids should be in column_ids?
```
3. Even if I pushed the IsNullFilter, The result.size() of DataTable::Scan() seems include all the rows. I traced into the TemplatedNullSelection(), and find the second parameter approved_tuple_count is passed as value, is it expected? Anyway, I changed it to a reference and set it to result size, seems it works for me. Please let me know if I mistake anything.

### CODE REVIEW

File: src/catalog/catalog_entry/table_catalog_entry.cpp [hunk 1]
```diff
@@ -380,6 +390,91 @@ unique_ptr<CatalogEntry> TableCatalogEntry::SetDefault(ClientContext &context, S
...
+
+		// filter
+		TableFilterSet table_filters;
+		table_filters.PushFilter(not_null_idx, make_unique<IsNullFilter>());
+
+		// state
+		TableScanState state;
+
+		storage->InitializeScan(transaction, state, column_ids, &table_filters);
+		storage->Scan(transaction, result, state, column_ids);
```
Comment (Mytherin) [inline src/catalog/catalog_entry/table_catalog_entry.cpp:439]: This only scans a single column - this will not work for larger tables.

File: src/storage/table/column_segment.cpp [hunk 2]
```diff
@@ -267,7 +267,7 @@ static void FilterSelectionSwitch(T *vec, T *predicate, SelectionVector &sel, id
 }
 
 template <bool IS_NULL>
-static idx_t TemplatedNullSelection(SelectionVector &sel, idx_t approved_tuple_count, ValidityMask &mask) {
+static idx_t TemplatedNullSelection(SelectionVector &sel, idx_t &approved_tuple_count, ValidityMask &mask) {
```
Comment (Mytherin) [inline src/storage/table/column_segment.cpp:270]: Why is this change necessary?

File: src/catalog/catalog_entry/table_catalog_entry.cpp (hunk 1 above)
Comment (Mytherin) [inline src/catalog/catalog_entry/table_catalog_entry.cpp:439]: I would prefer if this were generalized to add and drop arbitrary `CHECK` constraints - as most of the code should be able to be re-used.

File: test/sql/alter/alter_column/test_drop_not_null.test [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# name: test/sql/alter/not_null/test_drop_not_null.test
+# description: Test ALTER TABLE ALTER COLUMN column_name DROP NOT NULL
+# group: [alter_column]
+
+# Happy path
+
+statement ok
```
Comment (Mytherin) [inline test/sql/alter/alter_column/test_drop_not_null.test:7]: Could you add `PRAGMA enable_verification` to all your tests please

File: test/sql/alter/alter_column/test_drop_not_null.test [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+CREATE TABLE test(i INTEGER, j INTEGER NOT NULL)
```
Comment (Mytherin) [inline test/sql/alter/alter_column/test_drop_not_null.test:8]: Could you add the following tests:

* Tests with a larger table
* Tests with adding constraints to tables with uncommitted changes
* Rolling back the adding and removal of constraints
* Testing of adding constraints and removing in the presence of multiple transactions
* Testing of adding and removing constraints with the storage and the WAL
* Adding and dropping constraints within the same transaction

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good, comments:

Also please have look at the test failures. You can look at test failures on your own fork.

File: test/sql/alter/alter_column/test_drop_not_null.test (hunk 3 above)
Comment (zippond) [reply test/sql/alter/alter_column/test_drop_not_null.test:7]: Sure. Will do that. Thank!

File: src/storage/table/column_segment.cpp (hunk 2 above)
Comment (zippond) [reply src/storage/table/column_segment.cpp:270]: Actually this is question #3 I raised in PR description. 

> 3.Even if I pushed the IsNullFilter, The result.size() of DataTable::Scan() seems include all the rows. I traced into the TemplatedNullSelection(), and find the second parameter approved_tuple_count is passed as value, is it expected? Anyway, I changed it to a reference and set it to result size, seems it works for me. Please let me know if I mistake anything.

TemplatedNullSelection is called in
```
idx_t ColumnSegment::FilterSelection(SelectionVector &sel, Vector &result, const TableFilter &filter,
                                     idx_t &approved_tuple_count, ValidityMask &mask) {
```

The reference approved_tuple_count passed in is set by all filters except for IS_NULL and IS_NOT_NULL, is it on purpose?

File: src/catalog/catalog_entry/table_catalog_entry.cpp (hunk 1 above)
Comment (Mytherin) [reply src/catalog/catalog_entry/table_catalog_entry.cpp:439]: Apologies I meant to say this only scans a single vector. This will scan at most 1024 values. `storage->Scan` needs to be called in a loop.

File: src/storage/table/column_segment.cpp (hunk 2 above)
Comment (zippond) [reply src/storage/table/column_segment.cpp:270]: Actually this is question [#3] I raised in PR description.

> 3.Even if I pushed the IsNullFilter, The result.size() of DataTable::Scan() seems include all the rows. I traced into the TemplatedNullSelection(), and find the second parameter approved_tuple_count is passed as value, is it expected? Anyway, I changed it to a reference and set it to result size, seems it works for me. Please let me know if I mistake anything.

I assueme approved_tuple_count is count of qualified(is null) rows, shouldn't it?
TemplatedNullSelection is called in 
```
idx_t ColumnSegment::FilterSelection(SelectionVector &sel, Vector &result, const TableFilter &filter,
                                     idx_t &approved_tuple_count, ValidityMask &mask) {
```

The reference approved_tuple_count passed in is set by all filters except for IS_NULL and IS_NOT_NULL, I changed it and seems it fix my problem. Please let me know if I make a mistake here.

File: src/catalog/catalog_entry/table_catalog_entry.cpp (hunk 1 above)
Comment (zippond) [reply src/catalog/catalog_entry/table_catalog_entry.cpp:439]: Thanks. As this is my first PR and I might make many stupid mistakes. I think I can make the enhancement in future PRs.

Comment (zippond) [reply src/catalog/catalog_entry/table_catalog_entry.cpp:439]: Sorry, forget about my comment above, my test only works when append null value in the tail of the large table. You're right, I need to scan in a loop.

**Comment by zippond** [comment]
I have to close this PR for now as I don't have much time working on this, and there're still several significant issues to be addressed.
1. Scan with pushed-down IsNull filter is not support yet, so I have to use a full scan with an extra nullness check.
2. Transactional issues. Rollback not work well.
3. Alter column need a readlock. 
Many thanks to Mytherin. I'll send a new PR once above issues are correctly addressed.
